The test fails locally, for me:

<details><summary>vm-test-run-goupile.log</summary>
<p>

```shellSession
goupile: (finished: waiting for unit goupile.service, in 15.30 seconds)
goupile: waiting for TCP port 8889 on localhost
goupile # Connection to localhost (::1) 8889 port [tcp/ddi-tcp-2] succeeded!
goupile: (finished: waiting for TCP port 8889 on localhost, in 0.05 seconds)
goupile: must succeed: curl -q http://localhost:8889
goupile #   % Total    % Received % Xferd  Average Speed  Time    Time    Time   Current
goupile #                                  Dload  Upload  Total   Spent   Left   Speed
goupile #   0      0   0      0   0      0      0      0                              0[   14.789115] goupile[1369]: Init sandbox
goupile # [   14.790245] goupile[1369]: Init main database
goupile # [   14.859483] goupile[1369]: Migrate domain database: 0 to 117
goupile # [   14.924167] goupile[1369]: Init domain
goupile # [   14.926775] goupile[1369]: Warning: Using default 'goupile' name for domain
goupile # [   14.929808] goupile[1369]: Periodic timer set to 183.2 s
goupile # 100   4563   0   4563   0      0  20914      0                              0100   4563   0   4563   0      0  20143      0                              0100   4563   0   4563   0      0  19407      0                              0
goupile: (finished: must succeed: curl -q http://localhost:8889, in 0.29 seconds)
goupile: must succeed: curl -q http://goupile.local
goupile #   % Total    % Received % Xferd  Average Speed  Time    Time    Time   Current
goupile #                                  Dload  Upload  Total   Spent   Left   Speed
goupile #   0      0   0      0   0      0      0      0                              0100   4563   0   4563   0      0 138.0k      0                              0100   4563   0   4563   0      0 115.2k      0                              0100   4563   0   4563   0      0 99.27k      0
                 0
goupile: (finished: must succeed: curl -q http://goupile.local, in 0.08 seconds)
goupile: must succeed: curl -q http://localhost
goupile #   % Total    % Received % Xferd  Average Speed  Time    Time    Time   Current
goupile #                                  Dload  Upload  Total   Spent   Left   Speed
goupile #   0      0   0      0   0      0      0      0                              0100   4563   0   4563   0      0 202.2k      0                              0100   4563   0   4563   0      0 154.4k      0                              0100   4563   0   4563   0      0 127.3k      0
                 0
goupile: (finished: must succeed: curl -q http://localhost, in 0.07 seconds)
goupile: must succeed: run-goupile-test
goupile # [   15.168818] goupile-e2e[1443]: ================================================================================
goupile # [   15.171848] goupile-e2e[1443]: NOTE: The goupile e2e test can be run interactively either inside the vm or on the host
goupile # [   15.174467] goupile-e2e[1443]:   - First, run `nix-build -A nixosTests.goupile.driverInteractive` and `./result/bin/nixos-test-driver`
goupile # [   15.174900] goupile-e2e[1443]:   - Run `start_all()` inside the repl
goupile # [   15.176158] goupile-e2e[1443]:   - Then `$(nix-build -A nixosTests.goupile.interactive-script)/bin/run-goupile-test` to run the full test interactively
goupile # [   15.177199] goupile-e2e[1443]:   - Or `env PWDEBUG=1 $(nix-build -A nixosTests.goupile.interactive-script)/bin/run-goupile-test` to show the playwright inspector to debug
goupile # [   15.178334] goupile-e2e[1443]: ================================================================================
goupile # [   15.179138] goupile-e2e[1445]: Starting smoke test...
goupile # [   21.348301] goupile[1369]: Creating forced snapshot (previous one is old)
goupile # [   22.615161] goupile[1369]: ::1: Migrate instance database 'proj1.db': 0 to 147
goupile # [   23.034236] goupile[1369]: Exporting new FS views of 'proj1'
goupile # [  112.871740] goupile-e2e[1450]: Traceback (most recent call last):
goupile # [  112.873109] goupile-e2e[1450]:   File "/nix/store/sx0mi8d4msfwjln33zgn7j02idxhh9hv-basic_interaction_test.py", line 178, in <module>
goupile # [  112.873584] goupile-e2e[1450]:     run_test()
goupile # [  112.874295] goupile-e2e[1450]:     ~~~~~~~~^^
goupile # [  112.875031] goupile-e2e[1450]:   File "/nix/store/sx0mi8d4msfwjln33zgn7j02idxhh9hv-basic_interaction_test.py", line 54, in run_test
goupile # [  112.875748] goupile-e2e[1450]:     page.get_by_role("button", name="Assign").nth(1).click()
goupile # [  112.876877] goupile-e2e[1450]:     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
goupile # [  112.877907] goupile-e2e[1450]:   File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/sync_api/_generated.py", line 15631, in click
goupile # [  112.878937] goupile-e2e[1450]:     self._sync(
goupile # [  112.879989] goupile-e2e[1450]:     ~~~~~~~~~~^
goupile # [  112.880704] goupile-e2e[1450]:         self._impl_obj.click(
goupile # [  112.881674] goupile-e2e[1450]:         ^^^^^^^^^^^^^^^^^^^^^
goupile # [  112.882690] goupile-e2e[1450]:     ...<10 lines>...
goupile # [  112.883660] goupile-e2e[1450]:         )
goupile # [  112.884665] goupile-e2e[1450]:         ^
goupile # [  112.885355] goupile-e2e[1450]:     )
goupile # [  112.886112] goupile-e2e[1450]:     ^
goupile # [  112.886871] goupile-e2e[1450]:   File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/_impl/_sync_base.py", line 115, in _sync
goupile # [  112.887821] goupile-e2e[1450]:     return task.result()
goupile # [  112.888581] goupile-e2e[1450]:            ~~~~~~~~~~~^^
goupile # [  112.889313] goupile-e2e[1450]:   File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/_impl/_locator.py", line 162, in click
goupile # [  112.890056] goupile-e2e[1450]:     return await self._frame._click(self._selector, strict=True, **params)
goupile # [  112.891211] goupile-e2e[1450]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
goupile # [  112.892274] goupile-e2e[1450]:   File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/_impl/_frame.py", line 566, in _click
goupile # [  112.893048] goupile-e2e[1450]:     await self._channel.send("click", self._timeout, locals_to_params(locals()))
goupile # [  112.893763] goupile-e2e[1450]:   File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/_impl/_connection.py", line 69, in send
goupile # [  112.894744] goupile-e2e[1450]:     return await self._connection.wrap_api_call(
goupile # [  112.895724] goupile-e2e[1450]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
goupile # [  112.896694] goupile-e2e[1450]:     ...<3 lines>...
goupile # [  112.897686] goupile-e2e[1450]:     )
goupile # [  112.898651] goupile-e2e[1450]:     ^
goupile # [  112.899621] goupile-e2e[1450]:   File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/_impl/_connection.py", line 559, in wrap_api_call
goupile # [  112.900370] goupile-e2e[1450]:     raise rewrite_error(error, f"{parsed_st['apiName']}: {error}") from None
goupile # [  112.901130] goupile-e2e[1450]: playwright._impl._errors.TimeoutError: Locator.click: Timeout 90000ms exceeded.
goupile # [  112.901803] goupile-e2e[1450]: Call log:
goupile # [  112.902873] goupile-e2e[1450]:   - waiting for get_by_role("button", name="Assign").nth(1)
goupile: output: ================================================================================
NOTE: The goupile e2e test can be run interactively either inside the vm or on the host
  - First, run `nix-build -A nixosTests.goupile.driverInteractive` and `./result/bin/nixos-test-driver`
  - Run `start_all()` inside the repl
  - Then `$(nix-build -A nixosTests.goupile.interactive-script)/bin/run-goupile-test` to run the full test interactively
  - Or `env PWDEBUG=1 $(nix-build -A nixosTests.goupile.interactive-script)/bin/run-goupile-test` to show the playwright inspector to debug
================================================================================
Traceback (most recent call last):
  File "/nix/store/sx0mi8d4msfwjln33zgn7j02idxhh9hv-basic_interaction_test.py", line 178, in <module>
    run_test()
    ~~~~~~~~^^
  File "/nix/store/sx0mi8d4msfwjln33zgn7j02idxhh9hv-basic_interaction_test.py", line 54, in run_test
    page.get_by_role("button", name="Assign").nth(1).click()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/sync_api/_generated.py", line 15631, in click
    self._sync(
    ~~~~~~~~~~^
        self._impl_obj.click(
        ^^^^^^^^^^^^^^^^^^^^^
    ...<10 lines>...
        )
        ^
    )
    ^
  File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/_impl/_sync_base.py", line 115, in _sync
    return task.result()
           ~~~~~~~~~~~^^
  File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/_impl/_locator.py", line 162, in click
    return await self._frame._click(self._selector, strict=True, **params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/_impl/_frame.py", line 566, in _click
    await self._channel.send("click", self._timeout, locals_to_params(locals()))
  File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/_impl/_connection.py", line 69, in send
    return await self._connection.wrap_api_call(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<3 lines>...
    )
    ^
  File "/nix/store/n4g112y5fl0vbivyf7wmpsmqdnf1aw6h-python3-3.13.12-env/lib/python3.13/site-packages/playwright/_impl/_connection.py", line 559, in wrap_api_call
    raise rewrite_error(error, f"{parsed_st['apiName']}: {error}") from None
playwright._impl._errors.TimeoutError: Locator.click: Timeout 90000ms exceeded.
Call log:
  - waiting for get_by_role("button", name="Assign").nth(1)


!!! Traceback (most recent call last):
!!!   File "<string>", line 11, in <module>
!!!     machine.succeed("run-goupile-test")
!!!
!!! RequestedAssertionFailed: command `run-goupile-test` failed (exit code 1)
cleanup
kill QemuMachine (pid 11)
qemu-system-x86_64: terminating on signal 15 from pid 6 (/nix/store/pzdalg368npikvpq4ncz2saxnz19v53k-python3-3.13.12/bin/python3.13)
vde_switch: EOF data port: Interrupted system call
kill vlan (pid 7)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Caught signal 15, cleaning up and exiting
vde_switch: Could not remove ctl socket '/build/vde1.ctl': No such file or directory
(finished: cleanup, in 0.02 seconds)
```

</p>
</details>

--- a/nixos/modules/services/web-apps/goupile.nix
+++ b/nixos/modules/services/web-apps/goupile.nix
@@ -60,7 +60,8 @@ in
 
     hostName = lib.mkOption {
       type = lib.types.str;
-      default = "goupile";
+      default = config.networking.fqdnOrHostName;
+      defaultText = lib.literalExpression "config.networking.fqdnOrHostName";
       description = "Nginx service name for goupile service.";
     };
   };
@@ -70,7 +71,7 @@ in
         services.nginx = {
           enable = lib.mkDefault true;
           virtualHosts.${cfg.hostName} = {
-            locations."/".proxyPass = "http://${cfg.hostName}:${builtins.toString cfg.settings.HTTP.Port}";
+            locations."/".proxyPass = "http://127.0.0.1:${builtins.toString cfg.settings.HTTP.Port}";
           };
         };
       }

--- a/nixos/tests/web-apps/goupile/basic_interaction_test.py
+++ b/nixos/tests/web-apps/goupile/basic_interaction_test.py
@@ -1,13 +1,22 @@
 import os
 import openpyxl
 import tempfile
-from playwright.sync_api import sync_playwright
+from playwright.sync_api import sync_playwright, Page
 
 BASE_URL = "http://localhost:8889"
 
 # NOTE: these are the passwords
 ADMIN_PASSWD = "car-shop-in-the-mall"
 ALICE_PASSWD = "user-goes-to-the-car-shop"
+
+
+def click_save_and_wait(page: Page):
+    save_btn = page.get_by_role("button", name="Save")
+    with page.expect_response(
+        lambda r: r.request.method in ["POST", "PUT", "PATCH"] and r.status < 400
+    ):
+        save_btn.click()
+    page.wait_for_selector(".ui_busy", state="hidden", timeout=30000)
 
 
 def run_test():
@@ -20,6 +29,8 @@ def run_test():
         )
         # more default timeout for slow nixos test vms
         context.set_default_timeout(90 * 1000)
+        context.tracing.start(screenshots=True, snapshots=True, sources=True)
+
         page = context.new_page()
 
         page.goto(f"{BASE_URL}/admin")
@@ -48,10 +59,17 @@ def run_test():
         page.get_by_role("button", name="No", exact=True).click()
         page.get_by_role("textbox", name="Password *").fill(ALICE_PASSWD)
         page.get_by_role("textbox", name="Confirmation").fill(ALICE_PASSWD)
-        page.get_by_role("button", name="Create").click()
+
+        with page.expect_response(
+            lambda r: r.request.method == "POST" and r.status < 400
+        ):
+            page.get_by_role("button", name="Create").click()
 
         # give alice, permissions to access the project
-        page.get_by_role("button", name="Assign").nth(1).click()
+        assign_button = page.get_by_role("button", name="Assign").nth(1)
+        assign_button.wait_for(state="visible")
+        assign_button.click()
+
         page.get_by_text("Read", exact=True).click()
         page.get_by_text("Save", exact=True).click()
         page.get_by_text("Export", exact=True).click()
@@ -71,25 +89,23 @@ def run_test():
         page1.get_by_role("textbox", name="Inclusion date *").fill("2000-01-01")
         page1.get_by_role("spinbutton", name="Age *").click()
         page1.get_by_role("spinbutton", name="Age *").fill("1")
-        page1.get_by_role("button", name="Save").click()
-        page1.wait_for_timeout(1000)
+        click_save_and_wait(page1)
 
         page1.get_by_role("button", name="Advanced").click()
         page1.get_by_role("spinbutton", name="Age *").click()
         page1.get_by_role("spinbutton", name="Age *").fill("1")
-        page1.get_by_role("button", name="Save").click()
-        page1.wait_for_timeout(1000)
+        click_save_and_wait(page1)
 
         page1.get_by_role("button", name="Page layout").click()
         page1.get_by_role("spinbutton", name="Variable A1").fill("1")
-        page1.get_by_role("button", name="Save").click()
-        page1.wait_for_timeout(1000)
+        click_save_and_wait(page1)
 
         # create export #1
         page1.get_by_role("button", name="Data").click()
-        page1.wait_for_timeout(1000)
 
+        page1.get_by_role("button", name="Data exports").wait_for(state="visible")
         page1.get_by_role("button", name="Data exports").click()
+
         with page1.expect_download() as download_info:
             page1.get_by_role("button", name="Create export").click()
 
@@ -115,26 +131,24 @@ def run_test():
         page2.get_by_role("textbox", name="Inclusion date *").fill("2000-01-01")
         page2.get_by_role("spinbutton", name="Age *").click()
         page2.get_by_role("spinbutton", name="Age *").fill("2")
-        page2.get_by_role("button", name="Save").click()
-        page2.wait_for_timeout(1000)
+        click_save_and_wait(page2)
 
         page2.get_by_role("button", name="Advanced").click()
         page2.get_by_role("spinbutton", name="Age *").click()
         page2.get_by_role("spinbutton", name="Age *").fill("2")
-        page2.get_by_role("button", name="Save").click()
-        page2.wait_for_timeout(1000)
+        click_save_and_wait(page2)
 
         page2.get_by_role("button", name="Page layout").click()
         page2.get_by_role("spinbutton", name="Variable A1").click()
         page2.get_by_role("spinbutton", name="Variable A1").fill("2")
-        page2.get_by_role("button", name="Save").click()
-        page2.wait_for_timeout(1000)
+        click_save_and_wait(page2)
 
         # create export #2
         page2.get_by_role("button", name="Data").click()
-        page2.wait_for_timeout(1000)
 
+        page2.get_by_role("button", name="Data exports").wait_for(state="visible")
         page2.get_by_role("button", name="Data exports").click()
+
         page2.get_by_role("button", name="Previous exports").click()
 
         with page2.expect_download() as download1_info:
@@ -156,18 +170,19 @@ def run_test():
 
         print(f"exported all records to {save_path2}")
 
+        context.tracing.stop(path="/tmp/videos/trace.zip")
+
         context.close()
         browser.close()
 
     # check that exported files have correct entries
-
     wb1 = openpyxl.load_workbook(save_path1)
-    for sheet, cell in zip(["intro", "advanced", "layout"], ["D2", "D2", "C2"]):
+    for sheet, cell in zip(["intro", "advanced", "layout"], ["E2", "E2", "D2"]):
         val = wb1[sheet][cell].value
         assert val == 1, f"Sheet {sheet}, Cell {cell}: Expected 1 (admin), got {val}"
 
     wb2 = openpyxl.load_workbook(save_path2)
-    for sheet, cell in zip(["intro", "advanced", "layout"], ["D3", "D3", "C3"]):
+    for sheet, cell in zip(["intro", "advanced", "layout"], ["E3", "E3", "D3"]):
         val = wb2[sheet][cell].value
         assert val == 2, f"Sheet {sheet}, Cell {cell}: Expected 2 (alice), got {val}"
 

--- a/nixos/tests/web-apps/goupile/default.nix
+++ b/nixos/tests/web-apps/goupile/default.nix
@@ -73,7 +73,6 @@ in
       #systemd.services.goupile.environment.DEFAULT_SECCOMP_ACTION = "Log"; # Block|Log|Kill
       networking = {
         firewall.allowedTCPPorts = [ config.services.nginx.defaultHTTPListenPort ];
-        hostName = "goupile";
         domain = "local";
       };
 
@@ -108,7 +107,7 @@ in
       machine.wait_for_open_port(${port})
 
       machine.succeed("curl -q http://localhost:${port}")
-      machine.succeed("curl -q http://goupile.local")
+      machine.succeed("curl -q http://machine.local")
       machine.succeed("curl -q http://localhost")
 
       machine.succeed("run-goupile-test")

--- a/nixos/tests/web-apps/goupile/default.nix
+++ b/nixos/tests/web-apps/goupile/default.nix
@@ -110,9 +110,11 @@ in
       machine.succeed("curl -q http://machine.local")
       machine.succeed("curl -q http://localhost")
 
-      machine.succeed("run-goupile-test")
-      out_dir = os.environ.get("out", os.getcwd())
-      machine.copy_from_vm("/tmp/videos", out_dir)
+      try:
+          machine.succeed("run-goupile-test")
+      finally:
+          out_dir = os.environ.get("out", os.getcwd())
+          machine.copy_from_vm("/tmp/videos", out_dir)
     '';
 
   # Debug interactively with:

--- a/pkgs/by-name/go/goupile/package.nix
+++ b/pkgs/by-name/go/goupile/package.nix
@@ -27,14 +27,14 @@ let
 in
 stdenv'.mkDerivation (finalAttrs: {
   pname = "goupile";
-  version = "3.12.4";
+  version = "3.12.5";
 
   # https://github.com/Koromix/rygel/tags
   src = fetchFromGitHub {
     owner = "Koromix";
     repo = "rygel";
     tag = "goupile/${finalAttrs.version}";
-    hash = "sha256-2lHCOsvjTZeRkboefIdyh7JoSmes3KgvjFnXKnQ4On4=";
+    hash = "sha256-REgop8kMiB9j64hMsDmyuYidsKgMtKVYf/lwRbuf5Ho=";
   };
 
   nativeBuildInputs = [

